### PR TITLE
SYS-3369 add bridge_contract storage item and key functionality

### PR DESCRIPTION
**Pull request by MBrozhko34**
These changes consolidate the previously existing storage items 'ValidatorsManagerContract', 'LiftingContract' ( ethereum-events pallet) and 'PublishRootContract' (ethereum-transactions pallet) into one storage item in the avn pallet, 'EthereumBridgeContract', along with the relevant functionality.

### CODE REVIEW

File: pallets/ethereum-transactions/src/lib.rs
Comment (alexpArtos) [inline pallets/ethereum-transactions/src/lib.rs:212]: Do we need to leave this comment here?

**Review by alexpArtos — APPROVED** [review]

File: pallets/avn/src/lib.rs
Comment (ivan-cholakov) [inline pallets/avn/src/lib.rs:579]: did you check if you need this

File: pallets/ethereum-transactions/src/lib.rs
Comment (MBrozhko34) [reply pallets/ethereum-transactions/src/lib.rs:212]: Thanks for noticing, will remove

File: pallets/avn/src/lib.rs
Comment (MBrozhko34) [reply pallets/avn/src/lib.rs:579]: It seems within the other pallets it is more common practice to not have the mock.rs file within the /tests/ subfolder

**Review by ivan-cholakov — APPROVED** [review]

**Comment by ivan-cholakov** [comment]
@MBrozhko34  can you run and push `cargo fmt` ???

**Review by thadouk — CHANGES_REQUESTED** [review]
If migration is not included in this PR, then the upgrade path is broken.
We should avoid merging till that is implemented. Can we please change the target to a feature branch?

File: pallets/ethereum-transactions/src/tests/tests.rs
Comment (alexpArtos) [inline pallets/ethereum-transactions/src/tests/tests.rs:243]: Why? I do like this pipeline style, where each line shows a different operation in a clear and much more visible way.

Comment (alexpArtos) [reply pallets/ethereum-transactions/src/tests/tests.rs:243]: This is a personal preference, but I don't feel the need to save on number of lines and instead write longer lines.
We're in an age where we don't print code any more, so it's not about saving paper. In that sense, I don't find it a problem if files are longer, but I sure do find a problem if I have to scroll left and right, as well as up and down. 

And that said, even if I don't have to scroll left and right, it costs me mental energy if I have to parse even a short line to find where the breaking points are. If the code layout could tell me that immediately, it would save me energy and keep me productive for longer in the day.

Comment (MBrozhko34) [reply pallets/ethereum-transactions/src/tests/tests.rs:243]: I think this may be related to the code formatting linting checks as I have now removed the with_genesis_config(). Will double check.

File: pallets/avn/src/tests/test_proxy_signed_add_ethereum_logs.rs
Comment (MBrozhko34) [reply pallets/avn/src/tests/test_proxy_signed_add_ethereum_logs.rs:100]: The test for the nft contract is still in the ethereum-events pallet test_proxy_signed_add_ethereum_logs file. This test file is just testing for the new bridge contract.

File: pallets/ethereum-events/src/lib.rs
Comment (MBrozhko34) [reply pallets/ethereum-events/src/lib.rs:110]: This file name was changed from 'test_set_ethereum_contract' to 'test_map_nft_contract' which only performs tests for nft contract since the only remaining contract functionality within this pallet is for the nft contract.

File: pallets/ethereum-transactions/src/lib.rs
Comment (thadouk) [inline pallets/ethereum-transactions/src/lib.rs:206]: Do we need the genesis block?
We can eliminate it I think.

File: pallets/ethereum-transactions/src/lib.rs
Comment (thadouk) [inline pallets/ethereum-transactions/src/lib.rs:427]: Should we simplify this code and always return the bridge contract instead? We can also eliminate the function altogether if we want and just access the storage item from avn pallet.

File: pallets/ethereum-transactions/src/benchmarking.rs
Comment (thadouk) [reply pallets/ethereum-transactions/src/benchmarking.rs:147]: still needs removal :)

File: pallets/ethereum-events/src/mock.rs
Comment (thadouk) [inline pallets/ethereum-events/src/mock.rs:690]: Don't we need the avn pallet genesis here as well?

File: pallets/ethereum-events/src/mock.rs
Comment (thadouk) [inline pallets/ethereum-events/src/mock.rs:681]: is `with_genesis_config` run for all test configs? as is, it will be added in the genesis for the tests that use that, and not for example the `initial_lifts` one.

**Review by thadouk — APPROVED** [review]

File: pallets/avn/src/tests/test_proxy_signed_add_ethereum_logs.rs
Comment (alexpArtos) [reply pallets/avn/src/tests/test_proxy_signed_add_ethereum_logs.rs:52]: I don't think this touches the ValidatorsManager though. The Validator event is a placeholder, in fact, and I made a comment to change the function altogether not to use it. But there are multiple issues here.

One is, as you say, that in this file it no longer makes sense to test this function in particular since we only kept the nft contracts. That was not clear when I wrote the comment.

Another issue is that this function exists somewhere else, but no longer needs an argument. 

The other though is that in the context of fixing this test but not removing, only this part was removed. Inasmuch as the test still exists, removing the pre-condition feels wrong, and I wrote the comment to teach that part, this is, why we have these pre-conditions in place. Michael is still starting, so I use these initial PRs as talking points to explain general practices and, for example, test-writing philosophy.

File: pallets/ethereum-transactions/src/lib.rs
Comment (MBrozhko34) [reply pallets/ethereum-transactions/src/lib.rs:427]: I am not too sure - the enum 'EthTransactionType' has many different values, including 'Invalid' in which an empty address should be returned.

**Review by alexpArtos — CHANGES_REQUESTED** [review]

File: pallets/ethereum-events/src/benchmarking.rs
Comment (MBrozhko34) [reply pallets/ethereum-events/src/benchmarking.rs:258]: The function 'map_nft_contract' is on line 735 of ethereum-events/lib.rs. It is the modified function from 'set_ethereum_contract' that only deals with the nft contract in the ethereum-events as oppose to previously other contract types.

File: pallets/nft-manager/src/lib.rs
Comment (ivan-cholakov) [inline pallets/nft-manager/src/lib.rs:866]: isn't this part of another pr related to bounded vec? Maybe you should rebase and review your changes

File: pallets/parachain-staking/src/benchmarks.rs
Comment (ivan-cholakov) [inline pallets/parachain-staking/src/benchmarks.rs:1395]: This is also a part of another pr related to the boundedvec revert, so don't think this should be in your pr

**Comment by MBrozhko34** [comment]
Changed base to run checks, can change back to feature branch if required

File: pallets/avn/src/tests/test_proxy_signed_add_avn_logs.rs
Comment (alexpArtos) [inline pallets/avn/src/tests/test_proxy_signed_add_avn_logs.rs:56]: :+1:

File: pallets/ethereum-transactions/src/lib.rs
Comment (alexpArtos) [inline pallets/ethereum-transactions/src/lib.rs:708]: I realise we probably have added this contract from the other pallet, but just in case we haven't shouldn't we check if the root contract is in the new structure before we kill it?

Comment (alexpArtos) [reply pallets/ethereum-transactions/src/lib.rs:708]: Probably a nitpick :/

**Review by alexpArtos — APPROVED** [review]

**Comment by alexpArtos** [comment]
Well done, quite a comprehensive PR and task, that made you learn many different aspects of Substrate code.
:clap:

File: pallets/ethereum-events/src/lib.rs
Comment (alexpArtos) [inline pallets/ethereum-events/src/lib.rs:1696]:
```suggestion
        let lift_address = LiftingContractAddress::<T>::get();
```

File: pallets/ethereum-events/src/lib.rs
Comment (alexpArtos) [inline pallets/ethereum-events/src/lib.rs:1698]:
```suggestion
        let pub_root_address = PublishRootContract::<T>::get();
```

File: pallets/ethereum-events/src/lib.rs
Comment (alexpArtos) [inline pallets/ethereum-events/src/lib.rs:1701]:
```suggestion
        if !lift_address.is_zero() && !validators_address.is_zero() && !pub_root_address.is_zero() {
```

Comment (alexpArtos) [reply pallets/ethereum-events/src/lib.rs:1701]: Let's try to be consistent with variable names. We had 3 contract addresses, and each address followed a different convention:
- just `address` (could have been `address1, address2, address3`)
- properly qualified address (with `validators` indicating usage)
- old style short variable name, with letters _suggesting_ usage but not quite (`pb_address`). I know it's local variables, but no need to require a different mindset in every line.

**Review by alexpArtos — APPROVED** [review]

File: pallets/ethereum-events/src/lib.rs
Comment (alexpArtos) [reply pallets/ethereum-events/src/lib.rs:1701]: The other thing about this block is that it introduces a strong dependency between the 3 addresses. The data migration will only happen if all 3 addresses are defined.
I don't mind the code being like this if we check the state of the chain before migrating, since this is a one-off, but if any of the chains (from test to mainnet) turns out to have conditions that will stop this migration, then this has to be changed.

**Review by alexpArtos — APPROVED** [review]

**Review by alexpArtos — APPROVED** [review]

File: pallets/avn/src/tests/test_proxy_signed_add_avn_logs.rs
Comment (thadouk) [inline pallets/avn/src/tests/test_proxy_signed_add_avn_logs.rs:52]:
```suggestion
                let context = Context::default();
```

File: pallets/avn/src/tests/test_proxy_signed_add_avn_logs.rs
Comment (thadouk) [inline pallets/avn/src/tests/test_proxy_signed_add_avn_logs.rs:1]: I think there is a mixup here.
This file says it tests proxy signed add avn logs but it actually tests the set_bridge_contract method.
It also feels like a duplicated file.

File: pallets/avn/src/tests/test_set_bridge_contract.rs
Comment (thadouk) [inline pallets/avn/src/tests/test_set_bridge_contract.rs:52]:
```suggestion
                let context = Context::default();
```

File: pallets/ethereum-events/src/lib.rs
Comment (thadouk) [inline pallets/ethereum-events/src/lib.rs:295]: :+1:

File: pallets/ethereum-events/src/lib.rs
Comment (thadouk) [inline pallets/ethereum-events/src/lib.rs:809]: I think similar with some lines above, we should do a check if the storage version is already on v4 and if so, skip the migration.

File: pallets/ethereum-events/src/lib.rs
Comment (thadouk) [inline pallets/ethereum-events/src/lib.rs:1706]:
```suggestion
                consumed_weight.saturating_add(T::DbWeight::get().read_writes(1, 2));
```
Each of these blocks do one read and two writes.

File: pallets/ethereum-events/src/lib.rs
Comment (thadouk) [inline pallets/ethereum-events/src/lib.rs:1690]: I think this migration logic can be simplified by using a function to decide which value to use (lift_address, validator_manager_address or pub_root_address) and assign it to variable, say `avn_bridge_address_to_use` and a single block of code that does the update..
Maybe I am missing something so lets discuss offline.

File: pallets/ethereum-events/src/lib.rs
Comment (thadouk) [inline pallets/ethereum-events/src/lib.rs:1742]: Seems like this will always get logged, in successes and failures.

File: pallets/ethereum-events/src/tests/test_map_nft_contract.rs
Comment (thadouk) [inline pallets/ethereum-events/src/tests/test_map_nft_contract.rs:1]:
```suggestion
// Copyright 2023 Aventus (UK) Ltd.
```

File: pallets/ethereum-events/src/tests/tests.rs
Comment (thadouk) [inline pallets/ethereum-events/src/tests/tests.rs:11]:
```suggestion
```

File: pallets/ethereum-events/src/tests/tests.rs
Comment (thadouk) [inline pallets/ethereum-events/src/tests/tests.rs:15]:
```suggestion
```

File: pallets/ethereum-transactions/src/mock.rs
Comment (thadouk) [inline pallets/ethereum-transactions/src/mock.rs:258]: why do we need this? are we not using the `with_genesis_config`?

File: pallets/ethereum-transactions/src/tests/tests_validate_unsigned.rs
Comment (thadouk) [inline pallets/ethereum-transactions/src/tests/tests_validate_unsigned.rs:142]: we change these call not to use `with_genesis_config`, why?

**Review by thadouk — APPROVED** [review]
Approved provisionally, needs some of the feedback addressed before merging.
Also if possible please merge this after the ethereum-event bounded vector work, as this change will affect other parts of the ecosystem, tools and dapps and will need some time to reach production

**Review by fluorostani — APPROVED** [review]

File: pallets/avn/src/lib.rs
Comment (nahuseyoum) [inline pallets/avn/src/lib.rs:168]: shouldn't we emit an event here?

File: pallets/ethereum-events/Cargo.toml
Comment (nahuseyoum) [inline pallets/ethereum-events/Cargo.toml:32]: why are we adding this dependency?

File: pallets/ethereum-events/src/lib.rs
Comment (nahuseyoum) [inline pallets/ethereum-events/src/lib.rs:20]: is this only used for migration?

File: pallets/ethereum-events/src/lib.rs
Comment (nahuseyoum) [inline pallets/ethereum-events/src/lib.rs:200]: can we remove this?

File: pallets/ethereum-events/src/lib.rs
Comment (nahuseyoum) [inline pallets/ethereum-events/src/lib.rs:742]: not sure if the function name is describing the action. If all its doing is insert nft contract address then why not call it that?

File: pallets/ethereum-events/src/lib.rs
Comment (nahuseyoum) [inline pallets/ethereum-events/src/lib.rs:1672]: i would split this migration in 2 and only deal with val_mgr and lifting here. We can do a similar migration from eth transactions.

Comment (nahuseyoum) [reply pallets/ethereum-events/src/lib.rs:1672]: Unless we remove the pallet dependencies again after the migration has run.

File: pallets/avn/src/lib.rs
Comment (thadouk) [reply pallets/avn/src/lib.rs:168]: yeah we can do

File: pallets/ethereum-events/Cargo.toml
Comment (MBrozhko34) [reply pallets/ethereum-events/Cargo.toml:32]: The PublishRootContract (from ethereum-transactions) is also killed in the ethereum-events pallet. This functionality was done in response to a prior pr comment but can be changed if necessary.

File: pallets/ethereum-events/src/lib.rs
Comment (MBrozhko34) [reply pallets/ethereum-events/src/lib.rs:1672]: Thank you for the review, will do this

**Review by alexpArtos — APPROVED** [review]

**Comment by nahuseyoum** [comment]
@MikeBrozhko have you tested the storage migration script? I think you might need to add some code to the avn runtime to trigger it.

**Review by nahuseyoum — APPROVED** [review]

**Pull request merged**
